cmd/nsm-sriov-manager/nsm-sriov-manager.go

```diff
@@ -0,0 +1,418 @@
+		return err
+	}
+
+	for i, vf := range discoveredVFs.vfs {
+		logrus.Infof("VF: %s %+v", i, vf)
+		// Check if there is already vfio device for iommu group of VF
+		iommuGroup, err := strconv.Atoi(vf.iommuGroup)
+		if err != nil {
+			// Something wrong with iommu group for this VF
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			logrus.Errorf("fail to convert iommu group with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			continue
+		}
+		vfioDevice := fmt.Sprintf("/dev/vfio/%d", iommuGroup)
+		_, err = os.Lstat(vfioDevice)
+		if !os.IsNotExist(err) && err != nil {
+			// Something wrong with access vfio path for iommu group
+			// it is safer to skip it and remove from the list of available VFs
+			// on the host.
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			logrus.Errorf("fail to check for existing vfio device with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			continue
+		}
+		// vfio device for iommu group does not exist, need to create it
+		discoveredVFs.vfs[vf.pciAddr].VFIODevice = vfioDevice
+		if err := bindVF(vf); err != nil {
+			// Could not bind, cannot use, delete this VF
+			logrus.Errorf("fail to bind VF to vfio device with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			continue
+		}
+	}
+	return nil
+}
+
+func waitAndRetry(timeout time.Duration, retries int, check func() bool) error {
+	ticker := time.NewTicker(timeout)
+	for r := 0; r < retries; r++ {
+		select {
+		case <-ticker.C:
+			if check() {
+				return nil
+			}
+		}
+
+	}
+
+	return fmt.Errorf("timeout has expired")
+}
+
+func bindVF(vf *VF) error {
+	pciAddr := strings.Replace(vf.pciAddr, "-", ":", -1)
+	unbindPath := fmt.Sprintf("/sys/bus/pci/devices/%s/driver/unbind", pciAddr)
+	cmdUnbind := exec.Command("echo", pciAddr)
+	u, err := os.OpenFile(unbindPath, os.O_WRONLY, 0200)
+	if err != nil {
+		return fmt.Errorf("fail to open unbind path %s with error: %+v", unbindPath, err)
+	}
+	defer u.Close()
+	cmdUnbind.Stdout = u
+	if err := cmdUnbind.Run(); err != nil {
+		return fmt.Errorf("unbind failed with error: %+v", err)
+	}
+
+	bindArgs := fmt.Sprintf(" %s %s ", vf.pciVendor, vf.pciType)
+	cmdBind := exec.Command("echo", bindArgs)
+	bindPath := "/sys/bus/pci/drivers/vfio-pci/new_id"
+	b, err := os.OpenFile(bindPath, os.O_WRONLY, 0200)
+	if err != nil {
+		return fmt.Errorf("fail to open bind path %s with error: %+v", unbindPath, err)
+	}
+	defer b.Close()
+	cmdBind.Stdout = b
+	if err := cmdBind.Run(); err != nil {
+		return fmt.Errorf("bind failed with error: %+v", err)
+	}
+
+	return waitAndRetry(200*time.Millisecond, 5, func() bool {
+		_, err := os.Stat(vf.VFIODevice)
+		if err == nil {
+			return true
+		}
+		return false
+	})
+}
+
+func main() {
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+	logrus.Infof("Starting SRIOV Network Device Plugin...")
```

Comment:
Define 'SRIOV network device plugin'

Reply:
It is not a plugin, my bad forgot to edit it, thanks for catching it.